crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,922 @@
+        match self {
+            BlankLines::Zero => *other == 0,
+            BlankLines::Many { count, range: _ } => count.get() == *other,
+        }
+    }
+}
+
+impl PartialOrd<u32> for BlankLines {
+    fn partial_cmp(&self, other: &u32) -> Option<Ordering> {
+        self.count().partial_cmp(other)
+    }
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+enum Follows {
+    #[default]
+    Other,
+    Decorator,
+    Def,
+    Docstring,
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+enum Status {
+    /// Stores the indent level where the nesting started.
+    Inside(usize),
+    /// This is used to rectify a Inside switched to a Outside because of a dedented comment.
+    CommentAfter(usize),
+    #[default]
+    Outside,
+}
+
+/// Contains variables used for the linting of blank lines.
+#[derive(Debug, Default)]
+pub(crate) struct BlankLinesChecker {
+    follows: Follows,
+    fn_status: Status,
+    class_status: Status,
+    /// First line that is not a comment.
+    is_not_first_logical_line: bool,
+    /// Used for the fix in case a comment separates two non-comment logical lines to make the comment "stick"
+    /// to the second line instead of the first.
+    last_non_comment_line_end: TextSize,
+    previous_unindented_line_kind: Option<LogicalLineKind>,
+}
+
+impl BlankLinesChecker {
+    /// E301, E302, E303, E304, E305, E306
+    pub(crate) fn check_lines(
+        &mut self,
+        tokens: &[LexResult],
+        locator: &Locator,
+        stylist: &Stylist,
+        indent_width: IndentWidth,
+        diagnostics: &mut Vec<Diagnostic>,
+    ) {
+        let mut prev_indent_length: Option<usize> = None;
+        let line_preprocessor = LinePreprocessor::new(tokens, locator, indent_width);
+
+        for logical_line in line_preprocessor {
+            self.check_line(
+                &logical_line,
+                prev_indent_length,
+                locator,
+                stylist,
+                diagnostics,
+            );
+            if !logical_line.is_comment_only {
+                prev_indent_length = Some(logical_line.indent_length);
+            }
+        }
+    }
+
+    #[allow(clippy::nonminimal_bool)]
+    fn check_line(
+        &mut self,
+        line: &LogicalLineInfo,
+        prev_indent_length: Option<usize>,
+        locator: &Locator,
+        stylist: &Stylist,
+        diagnostics: &mut Vec<Diagnostic>,
+    ) {
+        match self.class_status {
+            Status::Inside(nesting_indent) => {
+                if line.indent_length <= nesting_indent {
+                    if line.is_comment_only {
+                        self.class_status = Status::CommentAfter(nesting_indent);
+                    } else {
+                        self.class_status = Status::Outside;
+                    }
+                }
+            }
+            Status::CommentAfter(indent) => {
+                if !line.is_comment_only {
+                    if line.indent_length > indent {
```

Comment:
Because we only enter `Status::CommentAfter` when encountering a comment that is "outside" the indented block,  `>` is correct. On the contrary, using `>=` will cause E306 false positives.

We have tests like the one below.

```python
# E301
class Class:

    def fn1():
        pass
    # comment
    def fn2():
        pass
# end
```